### ISSUE TITLE
Regression - Artifacts and performances

### ISSUE THREAD

**Issue by Zyko0** [hajimehoshi/ebiten#2676]
### Ebitengine Version

v2.6.0-alpha.4

### Operating System

- [X] Windows
- [ ] macOS
- [ ] Linux
- [ ] FreeBSD
- [ ] OpenBSD
- [ ] Android
- [ ] iOS
- [ ] Nintendo Switch
- [ ] Xbox
- [ ] Web Browsers

### Go Version (`go version`)

go1.18.3 windows/amd64

### What steps will reproduce the problem?

(opengl)

- (case 1) Run this program https://gist.github.com/Zyko0/185f7b625a46cfc4c1df34efad29a6fa produces black bars for me when I press 'R', there are other mouse positions, where it does

- (case 2) Measure FPS in vsync off (in another program, not necessarily this one)

### What is the expected result?

- (case 1) Proper tiling and rendering when moving the mouse and after pressing 'R' key

- (case 2) Same or almost equivalent FPS than before*

### What happens instead?

- (case 1) Black bars are showing sometimes, and at the mouse offset after pressing 'R' key

- (case 2) Worse FPS than before*

### Anything else you feel useful to add?

This is a single issue containing two things (as discussed on discord), because they both seem to have been introduced by the same commit* https://github.com/hajimehoshi/ebiten/commit/1c09ec5e44727a0c38b605552d93e4d470a128ab
This was originally observed on v2.6.0-alpha.4, then found it might come from this specific commit.

- (case 1) It's only after this commit that I'm seeing black bars on the gist I linked above.
- (case 2) Before this commit, I get 3200+ fps with vsync off on a project I'm currently developping, while I get to 1900+ fps right after this commit.

**Comment by hajimehoshi** [comment]
I couldn't reproduce the case 1 with the latest commit 569f4724754a54e1a99ee010cbfd4a45bb62826e on my MacBook Pro 2020.

https://github.com/hajimehoshi/ebiten/commit/1c09ec5e44727a0c38b605552d93e4d470a128ab might change the texture size of the images, but I think reverting this doesn't resolve the root cause.

**Comment by hajimehoshi** [comment]
With https://github.com/hajimehoshi/ebiten/commit/569f4724754a54e1a99ee010cbfd4a45bb62826e

```
Update count per frame: 1
Internal image sizes:
  3: (2048, 2048)
  4: (2048, 2048)
  6: (3360, 2100)
  7: (4096, 4096)
Graphics commands:
  draw-triangles: dst: 3 <- src: [7, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 6, blend: {src-color: 1, src-alpha: 1, dst-color: 0, dst-alpha: 0, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 4 <- src: [7, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 6, blend: {src-color: 1, src-alpha: 1, dst-color: 0, dst-alpha: 0, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 4 <- src: [7, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 1734, blend: {src-color: 1, src-alpha: 1, dst-color: 5, dst-alpha: 5, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 3 <- src: [4, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 6, blend: {src-color: 1, src-alpha: 1, dst-color: 5, dst-alpha: 5, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 3 <- src: [7, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 678, blend: {src-color: 1, src-alpha: 1, dst-color: 5, dst-alpha: 5, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 6 (screen) <- src: [7, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 6, blend: {src-color: 1, src-alpha: 1, dst-color: 0, dst-alpha: 0, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 6 (screen) <- src: [3, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 6, blend: {src-color: 1, src-alpha: 1, dst-color: 5, dst-alpha: 5, op-color: 0, op-alpha: 0}, even-odd: false
```

With https://github.com/hajimehoshi/ebiten/commit/569f4724754a54e1a99ee010cbfd4a45bb62826e and reverting https://github.com/hajimehoshi/ebiten/commit/1c09ec5e44727a0c38b605552d93e4d470a128ab

```
Update count per frame: 1
Internal image sizes:
  1: (1024, 1024)
  2: (4096, 4096)
  3: (2048, 2048)
  4: (2048, 2048)
  6: (3360, 2100)
Graphics commands:
  draw-triangles: dst: 3 <- src: [1, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 6, blend: {src-color: 1, src-alpha: 1, dst-color: 0, dst-alpha: 0, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 6, blend: {src-color: 1, src-alpha: 1, dst-color: 0, dst-alpha: 0, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 4 <- src: [2, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 1734, blend: {src-color: 1, src-alpha: 1, dst-color: 5, dst-alpha: 5, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 3 <- src: [4, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 6, blend: {src-color: 1, src-alpha: 1, dst-color: 5, dst-alpha: 5, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 3 <- src: [1, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 678, blend: {src-color: 1, src-alpha: 1, dst-color: 5, dst-alpha: 5, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 6 (screen) <- src: [1, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 6, blend: {src-color: 1, src-alpha: 1, dst-color: 0, dst-alpha: 0, op-color: 0, op-alpha: 0}, even-odd: false
  draw-triangles: dst: 6 (screen) <- src: [3, (nil), (nil), (nil)], num of dst regions: 1, num of indices: 6, blend: {src-color: 1, src-alpha: 1, dst-color: 5, dst-alpha: 5, op-color: 0, op-alpha: 0}, even-odd: false
```

So there are slight differences of source texture sizes and the number of textures. Unfortunately, I don't think there is a perfect solution to fix the issue, as the behavior heavily depends on graphics cards. Even if I could fix this for @Zyko0's machine, this could cause another issue for another machine.

**Comment by hajimehoshi** [comment]
Does this happen with DirectX (11)?

**Comment by hajimehoshi** [comment]
The current status is to wait for @Zyko0 minimizing the test case and fixing it for the FPS issue. Thanks!

**Comment by Zyko0** [comment]
I minimized (as much as I could) the case for black bars
I tried reproducing the FPS issue in another test file, but I couldn't so far

**Comment by hajimehoshi** [comment]
> I minimized (as much as I could) the case for black bars

Is this https://gist.github.com/Zyko0/185f7b625a46cfc4c1df34efad29a6fa?

**Comment by Zyko0** [comment]
Not sure it's related, but might https://github.com/hajimehoshi/ebiten/pull/2681/files
I want to test after this commit just to see (just a note to myself)

**Comment by quasilyte** [comment]
I bumped into this issue as well.

@Zyko0 helped me out in Discord.

I'll copy some parts of the discussion here for reference.
I'll also come back after my game is released on Steam (I can't spend much time on debugging right now as I'm crunching).

----

My original question:

> If I have a big texture for the background in form of the ebiten.Image (let's say 3000x3000 pixels), should I chunk it into several smaller images, like 640x640 chunks or whatever or it's totally fine to use large single images and render only a sub-image from them?

![image](https://github.com/hajimehoshi/ebiten/assets/6286655/c82641b0-3bc6-4cd8-aae8-9974f8ffb40d)

> I think I'm hitting a big GC pause when Ebitengine works with them.
As you can see on this trace graph, there is a certain spike that makes the game freeze for something like 40ms, which is noticeable (it looks like a 1-2 frames drop). Through trial and error, I found out that this issue goes away if I stop rendering the background. The background is a big image as described above.

The trace is (v2.5.4):
```
github.com/hajimehoshi/ebiten/v2/internal/restorable.(*Image).WritePixels:300
github.com/hajimehoshi/ebiten/v2/internal/restorable.(*Image).ClearPixels:271
github.com/hajimehoshi/ebiten/v2/internal/atlas.(*Image).dispose:659
github.com/hajimehoshi/ebiten/v2/internal/atlas.(*Image).moveTo:234
github.com/hajimehoshi/ebiten/v2/internal/atlas.(*Image).putOnSourceBackend:360
github.com/hajimehoshi/ebiten/v2/internal/atlas.putImagesOnSourceBackend:118
github.com/hajimehoshi/ebiten/v2/internal/atlas.BeginFrame:855
github.com/hajimehoshi/ebiten/v2/internal/buffered.BeginFrame:37
github.com/hajimehoshi/ebiten/v2/internal/ui.(*context).updateFrameImpl:107
github.com/hajimehoshi/ebiten/v2/internal/ui.(*context).updateFrame:73
```

> This pause only happens once after the texture is generated and rendered a few frames (!)
So it's not the first Draw call, but rather some nth call that causes this to happen, which is surprising.
I would expect it to hang during the first texture draw attempt.

Zyko responded with:

> Could you try go getting v2.5.0-alpha.12.0.20230228174701-7c0fbce0cfd8 and test again please?

And it did, in fact, solve my issue.

> As funny as it may sound, but I can't reproduce this issue on this version.

Zyko continued:

> What about with this commit 1c09ec5e44727a0c38b605552d93e4d470a128ab (just one commit after the one I linked you)

This version does make things go bad.

**Comment by hajimehoshi** [comment]
Would it be possible to dump memory profiling just before GC happens?

**Comment by quasilyte** [comment]
@hajimehoshi I'll come back with some info after the release, please stay tuned. :)
I'm interested to investigate it further, but it may require some extra time that I don't posses at the moment.

**Comment by hajimehoshi** [comment]
@Zyko0 @quasilyte Can we have a minimal example to reproduce the issue? I have not succeeded to reproduce this yet.

**Comment by hajimehoshi** [comment]
My current rough understanding is:

* In `internal/atlas`, there are two types for backends: source and destination.
* If an image is used as a rendering destination, this image is moved from a source to a destination.
* If an image is used as a source for a while, the image is moved from a destination to a source.
* This timing is calculated by `usedAsSourceCount` and `destinationCount`
* After the change 1c09ec5e44727a0c38b605552d93e4d470a128ab, the timing to count `destinationCount` up was changed. This variable is less likely updated. (only when the image becomes an independent image)
* Due to this, the image is now much more likely moved from a destination to a source.

**Comment by hajimehoshi** [comment]
@Zyko0 @quasilyte Could you try 1ae5e022b65492024cb8dbc365be652a7915c25a ? Thanks,

**Comment by Zyko0** [comment]
I can confirm that https://github.com/hajimehoshi/ebiten/commit/1ae5e022b65492024cb8dbc365be652a7915c25a solves it for me
https://github.com/hajimehoshi/ebiten/commit/98cb77d94f7368588aae39d2be928ab835faffa2 still does, but shows a few less FPS than the first one (measurable but not significant imo: 2500 vs 2800 on the project I've been testing)

**Comment by quasilyte** [comment]
I tried both 1ae5e022b65492024cb8dbc365be652a7915c25a and 98cb77d94f7368588aae39d2be928ab835faffa2
I still get a 2-3 frames freeze at some points.
This could mean that my cause is slightly different.
I'll try to come up with a minimal reproducer and open a new issue.

**Comment by hajimehoshi** [comment]
Thanks! I'll close this issue for now.
